### PR TITLE
[Connector Secrets] Add delete API endpoint

### PR DESCRIPTION
**Pull request by navarone-feekery**
## Changes

- Add DELETE endpoint for `/_connector/_secret/{id}`
- Add endpoint to `write_connector_secrets` cluster privilege

### CODE REVIEW

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/secrets/action/DeleteConnectorSecretAction.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.connector.secrets.action;
+
+import org.elasticsearch.action.ActionType;
+
+public class DeleteConnectorSecretAction extends ActionType<DeleteConnectorSecretResponse> {
```
Comment (jedrazb) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/secrets/action/DeleteConnectorSecretAction.java:12]: There was a refactor of `ActionType` in `x-pack` last week. I think we should adhere to the new standard. Take a look at this PR https://github.com/elastic/elasticsearch/pull/104821/files

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/secrets/action/DeleteConnectorSecretRequest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        out.writeString(id);
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException validationException = null;
+
+        if (Strings.isNullOrEmpty(id)) {
```
Comment (jedrazb) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/secrets/action/DeleteConnectorSecretRequest.java:48]: NIT: can we test for this edge case in yaml tests?

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/520_connector_secret_delete.yml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  - set: { id: id }
+  - match: { id: $id }
+  - do:
+      headers: { Authorization: "Basic ZW50c2VhcmNoLXVucHJpdmlsZWdlZDplbnRzZWFyY2gtdW5wcml2aWxlZ2VkLXVzZXI=" }  # unprivileged
+      connector_secret.get:
+        id: $id
+      catch: unauthorized
+
+---
+'Get connector secret - Missing secret id':
```
Comment (jedrazb) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/520_connector_secret_delete.yml:67]: can we also test that null/ "" empty id causes 400 validation error?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearch.java [hunk 4]
```diff
@@ -272,7 +275,8 @@ protected XPackLicenseState getLicenseState() {
             actionHandlers.addAll(
                 List.of(
                     new ActionHandler<>(GetConnectorSecretAction.INSTANCE, TransportGetConnectorSecretAction.class),
-                    new ActionHandler<>(PostConnectorSecretAction.INSTANCE, TransportPostConnectorSecretAction.class)
+                    new ActionHandler<>(PostConnectorSecretAction.INSTANCE, TransportPostConnectorSecretAction.class),
```
Comment (jedrazb) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearch.java:278]: NIT: can we have alphabetical ordering here?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/secrets/ConnectorSecretsIndexService.java [hunk 5]
```diff
@@ -93,4 +95,15 @@ public void createSecret(PostConnectorSecretRequest request, ActionListener<Post
             listener.onFailure(e);
         }
     }
+
+    public void deleteSecret(String id, ActionListener<DeleteConnectorSecretResponse> listener) {
+        clientWithOrigin.prepareDelete(CONNECTOR_SECRETS_INDEX_NAME, id)
```
Comment (jedrazb) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/secrets/ConnectorSecretsIndexService.java:100]: is there any scenario in which we want to wrap this in `try/except` ? The API logic is responsible for handling it's own exceptions, any unhandled exception here can fail the node

**Review by jedrazb — COMMENTED** [review]
LGTM! Left a couple of comments but nothing major 👍

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/520_connector_secret_delete.yml (hunk 3 above)
Comment (navarone-feekery) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/520_connector_secret_delete.yml]: If this value is null or empty the request can't actually go through, so apparently the null check is redundant. The only GET or DELETE requests that are supported are `/_connector/_secret/{id}`. Making id null or empty changes the request to `/_connector/_secret/` which throws a 405.

There's no inbuilt method for this (e.g. `catch: method_not_allowed` isn't supported in the codebase). We also don't check this for similar endpoints like delete connector (we do a [null check](https://github.com/elastic/elasticsearch/blob/666774a8659643c6e779647023f26a5f4e54cce8/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/DeleteConnectorAction.java#L55-L57) but we don't have a test for it in its [yaml test](resources/rest-api-spec/test/entsearch/320_connector_delete.yml), probably for the same reason I'm having trouble adding one).

Do you think we should just remove the redundant null check? Or just continue forward without a yaml test for this?

Comment (jedrazb) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/520_connector_secret_delete.yml]: In that case I think it's fine to go ahead without the yaml test 👍

**Review by jedrazb — APPROVED** [review]
🚢

**Pull request merged**